packages/contracts/contracts/universal/MultiSigWallet.sol

```diff
@@ -0,0 +1,601 @@
+    }
+
+    /**
+     * @notice Allows to remove an owner. Transaction has to be sent by wallet.
+     *
+     * @param _owner Address of owner.
+     */
+    function removeOwner(address _owner)
+        external
+        override
+        onlyWallet
+        ownerExists(_owner)
+    {
+        isOwner[_owner] = false;
+        // find & delete item
+        for (uint256 i = 0; i < owners.length - 1;) {
+            if (owners[i] == _owner) {
+                owners[i] = owners[owners.length - 1];
+                owners.pop();
+                break;
+            }
+
+            unchecked {
+                ++i;
+            }
+        }
+
+        if (numConfirmationsRequired > owners.length) {
+            _changeNumConfirmationRequirement(owners.length);
+        }
+        emit OwnerRemoval(_owner);
+    }
+
+    /**
+     * @notice Allows to replace an owner with a new owner. Transaction has to be sent by wallet.
+     *
+     * @param _owner    Address of owner to be replaced.
+     * @param _newOwner Address of new owner.
+     */
+    function replaceOwner(address _owner, address _newOwner)
+        external
+        override
+        onlyWallet
+        ownerExists(_owner)
+        ownerDoesNotExist(_newOwner)
+    {
+        for (uint256 i = 0; i < owners.length;) {
+            if (owners[i] == _owner) {
+                owners[i] = _newOwner;
+                break;
+            }
+
+            unchecked {
+                ++i;
+            }
+        }
+
+        isOwner[_owner] = false;
+        isOwner[_newOwner] = true;
+        emit OwnerRemoval(_owner);
+        emit OwnerAddition(_newOwner);
+    }
+
+    /**
+     * @notice Allows to change number of confirmations required.
+     *
+     * @param _numConfirmationsRequired Number of required confirmations.
+     */
+    function _changeNumConfirmationRequirement(uint256 _numConfirmationsRequired)
+        internal
+        onlyWallet
+        validNumConfirmations(owners.length, _numConfirmationsRequired)
+    {
+        numConfirmationsRequired = _numConfirmationsRequired;
+    }
+
+    /**
+     * @notice Allows to change the number of required confirmations. Transaction has to be sent by wallet.
+     *
+     * @param _numConfirmationsRequired Number of required confirmations.
+     */
+    function changeRequirement(uint256 _numConfirmationsRequired)
+        external
+        override
+    {
+        _changeNumConfirmationRequirement(_numConfirmationsRequired);
+        emit RequirementChange(_numConfirmationsRequired);
+    }
+
+    /**
+     * @notice Allows an owner to submit and confirm a transaction.
+     *
+     * @param _destination Transaction target address.
+     * @param _value       Transaction ether value.
+     * @param _data        Transaction data payload.
+     *
+     * @return Returns transaction ID.
+     */
+    function submitTransaction(address _destination, uint256 _value, bytes memory _data)
+        public
+        override
+        virtual
```

Comment:
Don't we need to check `ownerExists` here? If it's true, I think it would be better to separate `confirmTransaction` to `public` function `confirmTransaction` and `internal` function `_confirmTransaction`, and check `ownerExists` in `public` function, because it might check twice `ownerExists` when call `submitTransaction`.

Reply:
Good point. Applied.

Reply:
Modified ownerExists check based on information like sender to be done entirely in public functions